File: opus-mk-ios/Opus MK/Opus MK/Lib/OpeningHours.swift
import Foundation

enum OpeningHoursHelper {
    /// Mirrors opus-mk `isOpenNow` (Europe/Belgrade, ISO weekday 0=Mon).
    static func isOpenNow(
        _ openingHours: [OpeningHoursEntry]?,
        now: Date = Date()
    ) -> Bool {
        guard let openingHours, !openingHours.isEmpty else { return false }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "Europe/Belgrade") ?? .current

        let weekday = calendar.component(.weekday, from: now)
        let currentDay = weekday == 1 ? 6 : weekday - 2

        let hour = calendar.component(.hour, from: now)
        let minute = calendar.component(.minute, from: now)
        let currentMinutes = hour * 60 + minute

        guard let today = openingHours.first(where: { $0.dayOfWeek == currentDay }),
              !today.isClosed
        else { return false }

        let openMins = minutes(from: today.open)
        let closeMins = minutes(from: today.close)

        if closeMins > openMins {
            return currentMinutes >= openMins && currentMinutes < closeMins
        }
        return currentMinutes >= openMins || currentMinutes < closeMins
    }

    private static func minutes(from time: String) -> Int {
        let parts = time.split(separator: ":").compactMap { Int($0) }
        guard parts.count >= 2 else { return 0 }
        return parts[0] * 60 + parts[1]
    }
}

enum DayPeriod {
    static var label: String {
        let now = Date()
        var calendar = Calendar.current
        calendar.timeZone = TimeZone(identifier: "Europe/Belgrade") ?? .current
        let dayIndex = calendar.component(.weekday, from: now) - 1
        let days = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
        let day = days[dayIndex]
        let hour = calendar.component(.hour, from: now)
        if hour < 12 { return "\(day) morning" }
        if hour < 17 { return "\(day) afternoon" }
        return "\(day) evening"
    }
}

File: opus-native/src/lib/discover/opening-hours.ts
import type { OpeningHoursEntry } from '@/lib/discover/types';

/** Mirrors opus-mk `isOpenNow` (Europe/Belgrade, ISO weekday 0 = Monday). */
export function isOpenNow(
  openingHours?: OpeningHoursEntry[],
  now: Date = new Date(),
  timezone = 'Europe/Belgrade',
): boolean {
  if (!openingHours?.length) return false;

  const formatter = new Intl.DateTimeFormat('en-US', {
    timeZone: timezone,
    weekday: 'short',
    hour: 'numeric',
    minute: 'numeric',
    hour12: false,
  });

  const parts = formatter.formatToParts(now);
  const hourPart = parts.find((p) => p.type === 'hour')?.value ?? '00';
  const minutePart = parts.find((p) => p.type === 'minute')?.value ?? '00';
  const currentMinutes = parseInt(hourPart, 10) * 60 + parseInt(minutePart, 10);

  const jsDay = now.getDay();
  const currentDayOfWeek = jsDay === 0 ? 6 : jsDay - 1;
  const previousDayOfWeek = currentDayOfWeek === 0 ? 6 : currentDayOfWeek - 1;

  const getMinutesFromTime = (timeStr: string) => {
    const [h, m] = timeStr.split(':').map(Number);
    return h * 60 + m;
  };

  const todaySchedule = openingHours.find((h) => h.dayOfWeek === currentDayOfWeek);
  if (todaySchedule && !todaySchedule.isClosed) {
    const openMins = getMinutesFromTime(todaySchedule.open);
    const closeMins = getMinutesFromTime(todaySchedule.close);

    if (closeMins > openMins) {
      if (currentMinutes >= openMins && currentMinutes < closeMins) return true;
    } else if (currentMinutes >= openMins) {
      return true;
    }
  }

  const yesterdaySchedule = openingHours.find((h) => h.dayOfWeek === previousDayOfWeek);
  if (yesterdaySchedule && !yesterdaySchedule.isClosed) {
    const openMins = getMinutesFromTime(yesterdaySchedule.open);
    const closeMins = getMinutesFromTime(yesterdaySchedule.close);
    if (closeMins < openMins && currentMinutes < closeMins) return true;
  }

  return false;
}
